src/rust/engine/src/nodes.rs

```diff
@@ -1104,18 +1104,24 @@ impl WrappedNode for Task {
       Self::run_wrapped_node_helper(deps, &context, params, product, func, entry)
         .await
         .into();
-    let result_val = function_call_result
-      .as_error_value_ref()
-      .map_err(|e| e.into_failure())?;
-    let (new_level, message, new_artifacts) = if can_modify_workunit {
-      (
-        engine_aware::EngineAwareLevel::retrieve(&context.core.types, &result_val),
-        engine_aware::Message::retrieve(&context.core.types, &result_val),
-        engine_aware::Artifacts::retrieve(&context.core.types, &result_val)
-          .unwrap_or_else(Vec::new),
-      )
-    } else {
-      (None, None, Vec::new())
+    let (new_level, message, new_artifacts) = {
+      // Get the returned object or raised exception as a Value, and extract any engine-aware
+      // parameters.
+      // NB: this means specific exceptions can be marked as engine-aware parameters which can
```

Comment:
This is interesting but I am not sure it's desirable; if we do want to allow `Exceptions` to subclass the engine-aware classes, we should document that and have some tests that demonstrate an example of using it. I'm tempted to think that we shouldn't allow it. Exceptions and engine-aware classes are both mildly complicated system features, and trying to use them together seems like it will cause trouble for future developers trying to debug that code, for relatively little gain.

Reply:
I agree. The comment can be removed. I don't know if specifically disallowing it makes too much sense though.